flink_jobs_v2/Timelines/src/main/java/timelines/Timeline.java

```diff
@@ -461,74 +459,64 @@ public int opInt(int[][][] truthTable, int op, int a, int b) {
-    }
+            if (endDt.isBefore(startDay) || startDt.isAfter(endDay)) { //exclude periods out of the runDate range
+                continue;
+            }
+            if (startDt.isBefore(startDay)) { //if a periods start is before 00:00:00 , set the periods's start the 00:00:00
+                startDt = startDay;
+            }
+            if (endDt.isAfter(endDay)) { // if a periods end is after the 23:59:59 , set the period's end the 23:59:59
+                endDt = endDay;
+            }
 
-    public void createDownTimeline(ArrayList<String> downPeriod, int defaultDownStatus, DateTime runDateDt) throws ParseException {
-        DateTime endDay = Utils.createDate("yyyy-MM-dd'T'HH:mm:ss'Z'", runDateDt.toDate(), 23, 59, 59);
+            boolean addCeiling = false;
+            int endFloorStatus = -1;
+            if (ceiling == null || ceiling.isAfter(endDt.getMillis())) { //if a timestamp exists after the period then keep the initial timeline's status for that period to be used later
+                addCeiling = true;
+                DateTime endFloor = samples.floorKey(endDt);
+                endFloorStatus = this.samples.get(endFloor);
 
-        Timeline downTimeline = null;
-        if (downPeriod != null) {
+            }
+            if (floor == null || !floor.equals(startDt.getMillis())) { // if start period does not match with the initial timeline's timestamp then need to add the start to the timeline with the specified status
+                this.samples.put(startDt, intStatus);                       //else the timestamp already contained and in next step its status will be replaced
+            }
+            //if a timestamp exists after a period then this timestamp should be added with the initial status of that period taken from the initial timeline
+            
+            if (addCeiling && endDt.plusMinutes(1).isBefore(endDay)) {
+                this.samples.put(endDt.plusMinutes(1), endFloorStatus);
+            }
 
-            DateTime startDown = Utils.convertStringtoDate("yyyy-MM-dd'T'HH:mm:ss'Z'", downPeriod.get(0));
-            DateTime endDown = Utils.convertStringtoDate("yyyy-MM-dd'T'HH:mm:ss'Z'", downPeriod.get(1));
-            endDown = endDown.plusMinutes(1);
-            TreeMap<DateTime, Integer> downSample = new TreeMap();
-            downSample.put(startDown, defaultDownStatus);
-            if (!endDown.isAfter(endDay.getMillis())) {
-                downSample.put(endDown, -1);
+            for (DateTime dt : samples.keySet()) {
+                if (!dt.isBefore(startDt) && !dt.isAfter(endDt)) { //if timestamps exist between the period then replace the timestamps with the specified status
+                    this.samples.replace(dt, intStatus);
+                }
             }
-            // We have downtime declared
-            downTimeline = new Timeline();
-            downTimeline.insertDateTimeStamps(downSample, true);
+            this.optimize();
```

Comment:
Should fill...() method take responsibility of automatically optimising() the timeline?

Reply:
Changed to avoid automatic optimization. Optimizing timeline needs to be added to computations where necessary